app/models/manager_refresh/save_collection/helper.rb

```diff
@@ -45,14 +46,22 @@ def save_inventory_collection!(inventory_collection, association)
       ActiveRecord::Base.transaction do
         association.find_each do |record|
           index = inventory_collection.object_index_with_keys(unique_index_keys, record)
-          if unique_db_indexes.include?(index) # Include on Set is O(1)
+          if unique_db_primary_keys.include?(record.id) # Include on Set is O(1)
+            # When we have a complex :through relations, it can easily happen a same result is returned multiple times.
+            # In this case we log a warning and ask for relation to return a correct distinct result. We don't want to
+            # do SELECT DISTINCT by default, since it's slow.
+            _log.warn("An :association of #{inventory_collection} is not returning a DISTINCT result, please make sure"\
+                      "the SQL query returns DISTINCT results.")
```

Comment:
How much is this really an issue? Would we end up in duplicates? And why isnt this caught by `unique_db_indexes` below? If I understand correctly `unique_db_indexes` checks by `manager_ref` whereas you are now also checking by `.id`

My main concern is, that this warning will most likely be ignored by developers 😄 
If its a problem, then raise an exception...